install.sh

```diff
@@ -276,77 +279,171 @@ assert_usable_kernel() {
-    cat > /tmp/sandstorm-userns-test.c << __EOF__
-#define _GNU_SOURCE
-#include <sched.h>
-#include <sys/types.h>
-#include <unistd.h>
-#include <stdio.h>
-
-int main() {
-  /* We're trying to verify that UID namespaces work when not root, so make sure we're
-   * not root. */
-  if (getuid() == 0) {
-    /* Number here doesn't really matter, but 65534 is usually "nobody". */
-    if (setuid(65534) < 0) {
-      perror("setuid");
-      return 1;
-    }
-  }
+    USERNS_CLONE_AT_ALL="no"
+  fi
 
-  /* OK, let's see if we can create a UID namespace. */
-  if (unshare(CLONE_NEWUSER) < 0) {
-    /* Nope. */
-    perror("unshare");
-    return 1;
-  }
+  # Clean up after ourselves.
+  rm -f "$USERNS_TEST_PROGRAM"
+}
+
+assert_userns_clone_works_or_can_be_made_to_work() {
+  # The purpose of this function is to bail out early if people have no way
+  # to run Sandstorm and we can't help them via e.g. sysctl.
+
+  # If they don't have working unprivileged user namespaces, then we bail.
+  if [ "$USERNS_CLONE_AT_ALL" = "no" ] ; then
+    fail "Your kernel does not appear to be compiled with" \
+         "support for unprivileged user namespaces (CONFIG_USER_NS=y), or something else is" \
+         "preventing creation of user namespaces. This feature is critical for sandboxing." \
+         "Arch Linux is known to ship with a kernel that disables this feature; if you are" \
+         "using Arch, you will unfortunately need to compile your own kernel (see" \
+         "https://bugs.archlinux.org/task/36969). If you are not using Arch, and don't" \
+         "know why your system wouldn't have user namespaces, please file a bug against" \
+         "Sandstorm so we can figure out what happened."
+  fi
+
+  # We know that unprivileged user namespaces basically work. If they
+  # work, and the user doesn't need a sysctl set, then we are happy.
+  if [ "$USERNS_CLONE_UNPRIVILEGED_NEEDS_SYSCTL_SET" = "no" ] ; then
+    return
+  fi
+
+  # At this point, we're going to need this sysctl set.
+  #
+  # But that won't work under one of a few circumstances. Let's identify
+  # those so we know to bail out.
+  local SYSCTL_PROBABLY_WORKS="yes"
+
+  # If /proc/sys is mounted read-only, then they won't be able to run
+  # sysctl with our help.
+  #
+  # This applies to Docker containers and probably other sorts of
+  # containers, too.
+  if egrep -q '/proc/sys\s+proc\s+ro' /proc/mounts ; then
+    SYSCTL_PROBABLY_WORKS="no"
+  fi
+
+  # If this system doesn't have a sysctl.conf, then we don't know how
+  # to set the default value for the next reboot.
+  if [ ! -e /etc/sysctl.conf ] ; then
+    SYSCTL_PROBABLY_WORKS="no"
+  fi
 
-  return 0;
+  # OK, so they need the sysctl set.
+  #
+  # If, however, they _can't_ set the sysctl, make the install fail
+  # now.
+  if [ "$SYSCTL_PROBABLY_WORKS" = "no" ] ; then
+    echo "# sysctl -w kernel.unprivileged_userns_clone=1"
+    echo "# cat >> /etc/sysctl.conf << __EOF__"
```

Comment:
Using `#` as a prompt character strikes me as slightly confusing here because it's also used as a comment just a couple lines later.

Maybe we don't need to append the comment in the case where the user is adding it manually, so you could say:

    echo 'kernel.unprivileged_userns_clone = 1' >> /etc/sysctl.conf

And now since it's just two single-liners, you could merely indent them rather than putting a fake-prompt in front of them, which the user might otherwise confusedly try to include in the command. This strikes me as consistent with your preference about putting quotes around "help". :)